app/code/community/Bolt/Boltpay/Model/FeatureSwitch.php

```diff
@@ -50,12 +94,109 @@ public function updateFeatureSwitches()
     }
+
+    /**
+     * Read feature switches from database if we didn't it before
+     */
+    protected function readSwitches()
+    {
+        if (isset($this->switches)) {
+            return;
+        }
+        $this->switches = unserialize(Mage::getStoreConfig('payment/boltpay/featureSwitches'));
+    }
+
+    /**
+     * Get unique ID from cookie or generate it and save it in cookie and session
+     * @return string
+     */
+    protected function getUniqueUserId()
+    {
+        $boltFeatureSwitchId = Mage::getSingleton('core/cookie')->get(SELF::COOKIE_NAME);
+        if (!$boltFeatureSwitchId) {
+            $boltFeatureSwitchId = uniqid("BFS", true);
+            $boltFeatureSwitchId = Mage::getSingleton('core/cookie')->set(SELF::COOKIE_NAME, $boltFeatureSwitchId, true);
+        }
+        return $boltFeatureSwitchId;
+    }
+
+    /**
+     * This method returns if a feature switch is enabled for a user.
+     * The way this is computed is as follows:
+     *
+     * @param string $switchName
+     * @param int $rolloutPercentage
+     *
+     * @return bool
+     */
+    protected function isInBucket($switchName, $rolloutPercentage)
+    {
+        $boltFeatureSwitchId = $this->getUniqueUserId();
+        $saltedString = $boltFeatureSwitchId . '-' . $switchName;
+        $position = crc32($saltedString) % 100;
+
+        return $position < $rolloutPercentage;
+    }
+
+    /**
+     * Fet feature switch object (value, defaultValue, rolloutPercentage) by switch name
+     *
+     * @param $switchName string
+     * @return object
+     * @throws Exception if default value for switch doesn't exists
+     */
+    protected function getFeatureSwitchValueByName($switchName){
+        $this->readSwitches();
+        if (!isset($this->defaultSwitches[$switchName])) {
+            throw new \Exception('Unknown feature switch');
+        }
+        if (isset($this->switches[$switchName])) {
+            return $this->switches[$switchName];
+        }
+        return $this->defaultSwitches[$switchName];
+    }
+
+    /**
+     * This returns if the switch is enabled.
+     *
+     * @param string $switchName name of the switch
+     *
+     * @return bool
+     * @throws \Exception
+     */
+    protected function isSwitchEnabled($switchName)
+    {
+        $switch = $this->getFeatureSwitchValueByName($switchName);
+        switch ($switch->rolloutPercentage) {
+            case 0:
+                return $switch->defaultValue;
+            case 100:
+                return $switch->value;
+            default:
+                $is_in_bucket = $this->isInBucket($switchName, $switch->rolloutPercentage);
+                return $is_in_bucket ? $switch->value : $switch->defaultValue;
+        }
+    }
+
+    /***************************************************
+     * Switch Helpers below
+     ***************************************************/
+    public function isSampleSwitchEnabled()
```

Comment:
Let's write a simple test for these two functions. Just asserting that isSwitchEnabled is called with the correct parameters is fine.

Reply:
Done